Feature: Add missing dhis2 periods

nshandra opened the pull request:
### :pushpin: References

* **Issue:** Closes [[User feedback] Add Six-monthly period for dataset](https://app.clickup.com/t/869b9v6pp)

### :memo: Implementation

Added  the following dataSet periods:
- `WeeklyWednesday`
- `WeeklyThursday`
- `WeeklySaturday`
- `WeeklySunday`
- `BiWeekly`
- `BiMonthly`
- `QuarterlyNov`
- `SixMonthly`
- `SixMonthlyApril`
- `SixMonthlyNov`

Added tests for all periods.

### :fire: Notes for the reviewer

### :video_camera: Screenshots/Screen capture

### :bookmark_tabs: Others
#869b9v6pp

nshandra committed:
feat: Add weekly period variants and fix weekly period not using isoWeeks. Add tests for existing period types

nshandra committed:
feat: add SixMonthly period and test

nshandra committed:
feat: add SixMonthlyApril and Nov and tests

nshandra committed:
feat: add BiWeekly and Nov and tests

nshandra committed:
feat: add BiMonthly and Nov and tests

nshandra committed:
feat: add QuarterlyNov and Nov and tests

ifoche commented:
Task linked: [CU-869b9v6pp [User feedback] Add Six-monthly period for dataset](https://app.clickup.com/t/869b9v6pp)

nshandra committed:
feat: add Financial periods

nshandra committed:
feat: switch periods.js to ts.

xurxodev reviewed (changes requested):
thanks @nshandra 

This PR expands the supported DHIS2 dataset period types and adds a new TypeScript implementation of `buildAllPossiblePeriods`, along with Jest coverage for all newly added variants. 

It's great.

I've found some issues to review.

## 1. Should fix

-   `src/domain/entities/DataForm.ts` declares `periodType?: DataFormPeriod`, but `src/webapp/utils/periods.ts` can throw for missing/unsupported `periodType`.
    `buildAllPossiblePeriods` should not throw when `periodType` is missing; either accept `periodType?: DataFormPeriod` and return `[]`, or add an early guard in the function.
-   Improve type safety/readability in `src/webapp/utils/periods.ts` and push it towards a more functional/immutable style without introducing mutable `let` state. It's safe this change because now we have unit test. Examples:
    -   eliminate `let current;` in `generateSixMonthlyNovPeriods` by computing `current` as a `const` expression (ternary) and, where feasible, use `clone().add(...)` to avoid mutating Moment instances;
    -   type the accumulator/output arrays explicitly (e.g. `const dates: string[] = []`) so the contract is clear at a glance;
    -   make intermediate values explicit/typed in helpers (so the types of `start/end/current` and formatting pieces are obvious during review).

## 2. Recommendations non blocking

-   Tests status: `src/test/utils/periods.spec.ts` covers the newly added period variants and passes (`yarn test --runInBand src/test/utils/periods.spec.ts`). Consider adding a couple more edge cases (e.g. `startDate > endDate`) to validate robustness.

nshandra committed:
Merge branch 'development' into feat/add-missing-dhis2-periods

nshandra committed:
feat: add input validation tests, handle undefined periodType.

nshandra committed:
refactor: avoid variable mutation, add typing.

nshandra committed:
feat: add leap day and ISO weeks related tests

nshandra committed:
feat: add DataFormPeriod type guard, use guard in download page

nshandra committed:
feat: add translations of onPeriodValidationError message. Avoid reporting undefined, use unknown instead.

nshandra commented:
@xurxodev 

Added:
- Period input validation tests
- Leap day and ISO weeks tests
- `DataFormPeriod` guard that removes template selection if the template has an incompatible periodType.

Refactor:
- improve type safety and inmutability in `src/webapp/utils/periods.ts`

xurxodev reviewed (changes requested):
thanks @nshandra                                                                                                                                                                                                         
                                                                                                                                                                                                                           
  The following items emerged from the commits that addressed the first review round.                                                                                                                                      
   
  1. Should fix                                                                                                                                                                                                            
                  
  1.1. FinancialType and WeeklyPeriodType duplicate DataFormPeriod subsets                                                                                                                                                 
   
  src/webapp/utils/periods.ts:233 defines type FinancialType = "FinancialApril" | "FinancialJuly" | "FinancialOct" | "FinancialNov" and WeeklyPeriodType at line 77. These are subsets of DataFormPeriod but defined       
  independently — if a new financial or weekly period is added to dataFormPeriods, these local types won't update. Extract them from DataFormPeriod using Extract<>:
                                                                                                                                                                                                                           
  type FinancialType = Extract<DataFormPeriod, `Financial${string}`>;
  type WeeklyPeriodType = Extract<DataFormPeriod, "Weekly" | `Weekly${string}`>;                                                                                                                                           
                                                                                                                                                                                                                           
  This ensures they stay in sync with the source of truth.                                                                                                                                                                 
                                                                                                                                                                                                                           
  1.2. isDataFormPeriod type guard uses any                                                                                                                                                                                
                  
  src/domain/entities/DataForm.ts:47 — the parameter type should be unknown instead of any:                                                                                                                                
   
  export function isDataFormPeriod(value: unknown): value is DataFormPeriod {                                                                                                                                              
      return typeof value === "string" && (dataFormPeriods as readonly string[]).includes(value);
  }                                                                                                                                                                                                                        
   
  Using any defeats the purpose of a type guard — callers can pass anything without the compiler flagging it.                                                                                                              
                  
  1.3. generateFinancialPeriods uses moment().month(monthName) — implicit locale dependency                                                                                                                                
                  
  src/webapp/utils/periods.ts:237 — moment().month(monthName).month() parses month names using the current locale. If the app locale is set to something other than English, "April" won't resolve to month index 3 and    
  periods will be silently wrong. Use an explicit map instead:
                                                                                                                                                                                                                           
  const financialMonthMap: Record<string, number> = { April: 3, July: 6, Oct: 9, Nov: 10 };
  const startMonthIndex = financialMonthMap[monthName];                                                                                                                                                                    
   
  2. Recommendations non blocking                                                                                                                                                                                          
                  
  2.1. onPeriodValidationError is optional — silent failure for other consumers                                                                                                                                            
   
  src/webapp/components/template-selector/TemplateSelector.tsx:54 — onPeriodValidationError? is optional. If a future consumer of TemplateSelector doesn't pass it, the unsupported period selection will silently reset   
  state without any user feedback. Consider either making it required or adding a fallback (e.g. console.warn) inside the component.
                                                                                                                                                                                                                           
  2.2. Test import uses @jest/globals but project uses Vitest

  src/test/utils/periods.spec.ts:2 — import { describe, it, expect } from "@jest/globals". The project migrated to Vitest. Consider importing from vitest instead to avoid confusion:                                      
   
  import { describe, it, expect } from "vitest";

nshandra committed:
fix: brittle types and locale dependent code.

nshandra committed:
fix: remove any from isDataFormPeriod

nshandra committed:
fix: make onPeriodValidationError non-optional

nshandra committed:
Merge branch 'development' into feat/add-missing-dhis2-periods

nshandra committed:
fix: remove jest import from periods test

nshandra commented:
@xurxodev 

### 1. To Fix
- 1.1. Fixed the sub-types to be extracted from `DataFormPeriod`
- 1.2. Fixed.
- 1.3. AFAIK yo have to set moment locale via `moment.locale('..')` but i agree that it can lead to hard to debug issues. Changed.

### 2. Recommended

- 2.1. Given than `TemplateSelector` is only used in `DownloadTemplatePage` i made it required.
- 2.2. I see the branch feature/security has the switch to Vitest but its not merged yet. Import removed as the test works without it.

xurxodev reviewed (approved):
thanks @nshandra

Ramon-Jimenez reviewed (changes requested):
@nshandra About this one. I am not sure if it is related to this PR, but in any case let's fix it.

I created on DEV EST several datasets for same OrgUnit, one for each DHIS2 period type implemented here. Thing is, when populating template with data for one dataset, it brings you data for all datasets and all periods, mixing ones with the others. In the screenshot, the selected dataset is BiMonthly, but along with a couple of BiMonthly Periods, it is also retrieving all other different DHIS2 Periods I configured for testing, which correspond to other datasets.

<img width="1486" height="575" alt="Image" src="https://github.com/user-attachments/assets/3928998d-6834-4ffb-9540-2c8a1e9ef9fb" />

I would expect that it only pulls the periods associated with the dataset selected before the spreadsheet generation.

Thanks!

Ramon-Jimenez committed:
Merge branch 'development' into feat/add-missing-dhis2-periods
